### DISCUSSION TITLE
Building on Linux

### DISCUSSION THREAD

**Discussion by tig** [tui-cs/Terminal.Gui#1229] [Q&A]
I must be holding it wrong.

I have both Ubuntu 24.04 and Debian 10 installed in WSL.

On both, `dotnet build` or `dotnet build -c Debug` works fine.

However, on both, `dotnet build -c Release` gives the following error.

![image](https://user-images.githubusercontent.com/585482/115257318-d4510800-a0e4-11eb-8600-4c7f2108430b.png)

Anyone have any idea what's going on?

**Comment by BDisp** [comment]
Maybe your Linux version (Debian) but I doubt. It's not affecting only to me, because others are already say the same. It's only happens using he `CursesDriver`.
I already tried to resolve but without success. I only run through SSH with remote access to shared folders. I don't build inside Linux.

**Reply by tznind** [reply]
What does list sdks show?

```bash
$ dotnet --list-sdks
3.1.407 [/usr/share/dotnet/sdk]
5.0.202 [/usr/share/dotnet/sdk]
```

I am able to build both Release and Debug on Debian

```
 lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 10 (buster)
Release:	10
Codename:	buster
```

```bash
thomas@mx:~/guiMain/gui.cs
$ dotnet clean
Microsoft (R) Build Engine version 16.9.0+57a23d249 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 19/04/2021 20:03:00.
     1>Project "/home/thomas/guiMain/gui.cs/Terminal.sln" on node 1 (Clean target(s)).
     1>ValidateSolutionConfiguration:
         Building solution configuration "Debug|Any CPU".
     1>Done Building Project "/home/thomas/guiMain/gui.cs/Terminal.sln" (Clean target(s)).

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.12
thomas@mx:~/guiMain/gui.cs
$ dotnet build -c Release
Microsoft (R) Build Engine version 16.9.0+57a23d249 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  Terminal.Gui -> /home/thomas/guiMain/gui.cs/Terminal.Gui/bin/Release/netstandard2.0/Terminal.Gui.dll
  Terminal.Gui -> /home/thomas/guiMain/gui.cs/Terminal.Gui/bin/Release/net5.0/Terminal.Gui.dll
  ReactiveExample -> /home/thomas/guiMain/gui.cs/ReactiveExample/bin/Release/netcoreapp3.1/ReactiveExample.dll
  Terminal.Gui -> /home/thomas/guiMain/gui.cs/Terminal.Gui/bin/Release/net472/Terminal.Gui.dll
  Example -> /home/thomas/guiMain/gui.cs/Example/bin/Release/net5.0/Example.dll
  UICatalog -> /home/thomas/guiMain/gui.cs/UICatalog/bin/Release/net5.0/UICatalog.dll
  UnitTests -> /home/thomas/guiMain/gui.cs/UnitTests/bin/Release/net5.0/UnitTests.dll
  Successfully created package '/home/thomas/guiMain/gui.cs/Terminal.Gui/bin/Release/Terminal.Gui.1.0.0-beta.10.32.nupkg'.
  Successfully created package '/home/thomas/guiMain/gui.cs/Terminal.Gui/bin/Release/Terminal.Gui.1.0.0-beta.10.32.snupkg'.

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:03.59

```

**Comment by tig** [comment]
```
3.1.404 [/usr/share/dotnet/sdk]
5.0.202 [/usr/share/dotnet/sdk]
```

Now it's building fine with either `-c Debug` or `-c Release` (I didn't change anything).

Weird.

**Reply by tznind** [reply]
Strange, maybe some kind of building race condition? or stale obj files or something.  I know when you build a root sln then building/testing  runs on all csproj in parallel (which can be an issue for state based tests - not that we have any of those).  Maybe it's something like that?

Can you set it to clean and build in a loop for a while see if it is reproducible?

**Reply by BDisp** [reply]
> Now it's building fine with either `-c Debug` or `-c Release` (I didn't change anything).
> 
> Weird.

Probably you had the `Terminal.Gui` open in VS2019 when you compiled on `Linux`. So, we have to close the project from the VS2019 in `Windows` first, before compile on `Linux`.

**Comment by BDisp** [comment]
I did a `dotnet clean` from `WSL` and I also having the same issue as @tig, because we are using the `Windows` volume.
Then I did `dotnet build -c Debug` without error and then `dotnet build -c Release` also without error.
Then I did a `dotnet clean` again and I haven't no errors this time.
How can I run the unit test from `WSL`, because I don't have errors on `Windows` but in the PR it seems to have errors? Thanks.

**Reply by BDisp** [reply]
Ok I confess it's the first time I run unit tests on `Linux` and is with `dotnet test` command. What I don't understand is why I get this error on `Linux` and not in `Windows` in the PR https://github.com/migueldeicaza/gui.cs/pull/1237:

```cs
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:01.01]     Terminal.Gui.ApplicationTests.KeyUp_Event [FAIL]
  Failed Terminal.Gui.ApplicationTests.KeyUp_Event [4 ms]
  Error Message:
   Assert.Equal() Failure
Expected: 8
Actual:   6
at Terminal.Gui.ApplicationTests.KeyUp_Event() in /gui.cs/UnitTests/ApplicationTests.cs:line 199
Failed!  - Failed:     1, Passed:   277, Skipped:     0, Total:   278, Duration: 1 s
```

**Reply by BDisp** [reply]
Well, I discovered that doing a `dotnet clean` in `Linux` at a pre compiled on `Windows` we always will have errors. So, we need to compile on `Linux` first if we need do to a clean.
I also discovered the unit test error and is because we can't use numbers variables on the left side of an `Assert.Equal/NotEqual`, it must be literal like `Assert.Equal(6, variable)` on `Linux` only.

```cs
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:   278, Skipped:     0, Total:   278, Duration: 1 s - /gui.cs/UnitTests/bin/Debug/net5.0/UnitTests.dll (net5.0)
```

**Reply by tig** [reply]
> I did a `dotnet clean` from `WSL` and I also having the same issue as @tig, because we are using the `Windows` volume.
> Then I did `dotnet build -c Debug` without error and then `dotnet build -c Release` also without error.
> Then I did a `dotnet clean` again and I haven't no errors this time.
> How can I run the unit test from `WSL`, because I don't have errors on `Windows` but in the PR it seems to have errors? Thanks.

`dotnet test`

the actual command the CI/CD uses is:  `dotnet test --no-restore --verbosity normal`
